engine/table/src/main/java/io/deephaven/engine/table/impl/sources/ring/AbstractRingChunkSource.java

```diff
@@ -0,0 +1,389 @@
+            }
+            final int ringIx = keyToRingIndex(key);
+            fillKey(destination, destOffset, ringIx);
+            ++destOffset;
+        }
+
+        @Override
+        public void close() {
+            destination.setSize(destOffset);
+        }
+    }
+
+    private final class RangeFiller implements LongRangeConsumer, Closeable {
+        private final WritableChunk<? super Values> destination;
+        private int destOffset;
+
+        public RangeFiller(WritableChunk<? super Values> destination) {
+            this.destination = Objects.requireNonNull(destination);
+        }
+
+        @Override
+        public void accept(long firstKey, long lastKey) {
+            // this check should not be necessary given precondition on fillChunk
+            if (!containsRange(firstKey, lastKey)) {
+                throw new IllegalStateException(String.format(
+                        "fillChunk precondition broken, invalid range. requested=[%d, %d], available=[%d, %d]",
+                        firstKey, lastKey, firstKey(), lastKey()));
+            }
+            final int firstRingIx = keyToRingIndex(firstKey);
+            final int lastRingIx = keyToRingIndex(lastKey);
+            final int size = fillByCopy(destination, destOffset, firstRingIx, lastRingIx);
+            if (size != lastKey - firstKey + 1) {
+                throw new IllegalStateException();
+            }
+            destOffset += size;
+        }
+
+        @Override
+        public void close() {
+            destination.setSize(destOffset);
+        }
+    }
+
+    private WritableChunk<Values> ring(int offset, int length) {
+        return ringView.resetFromArray(ring, offset, length);
+    }
+
+    private int fillByCopy(@NotNull WritableChunk<? super Values> destination, int destOffset, int firstRingIx,
+            int lastRingIx) {
+        // Precondition: valid firstRingIx, lastRingIx
+        if (firstRingIx <= lastRingIx) {
+            // Optimization when we can accomplish with single copy
+            final int size = lastRingIx - firstRingIx + 1;
+            destination.copyFromArray(ring, firstRingIx, destOffset, size);
+            return size;
+        }
+        return fillByCopy2(destination, destOffset, firstRingIx, lastRingIx);
+    }
+
+    private int fillByCopy2(@NotNull WritableChunk<? super Values> destination, int destOffset, int firstRingIx,
+            int lastRingIx) {
+        // Precondition: valid firstRingIx, lastRingIx
+        // Precondition: firstRingIx > lastRingIx
+        final int fillSize1 = capacity - firstRingIx;
+        final int fillSize2 = lastRingIx + 1;
+        destination.copyFromArray(ring, firstRingIx, destOffset, fillSize1);
+        destination.copyFromArray(ring, 0, destOffset + fillSize1, fillSize2);
+        return fillSize1 + fillSize2;
+    }
+
+    final int keyToRingIndex(long key) {
+        return (int) (key % capacity);
```

Comment:
Are you sure you don't want to use a power-of-two capacity requirement to make this calc more efficient?

Reply:
I don't trust JIT to optimize something like this if there are any instances with non-power-of-2 capacity.
I'm not sure about your tail comment; the tail costs a bit of extra row set math once per cycle, vs. the mod that is a tax on every cell access.
That said, this is not necessarily the most important thing to optimize.